From 8ac0f14af85f4cba4a1c3f995e1d46ca6598cdbd Mon Sep 17 00:00:00 2001
From: Jurijs Soloveckis <jsoloveckis@maxlinear.com>
Date: Fri, 2 Jun 2023 08:23:41 +0200
Subject: [PATCH] WLANRTSYS-55807: Implement new API: update_wan_metrics

    the command to update the wan metrics from the hostapd_cli:
    "update_wan_metrics
    <WAN Info>:<DL Speed>:<UL Speed>:<DL Load>:<UL Load>:<LMD>"
---
 hostapd/config_file.c |  2 +-
 hostapd/config_file.h |  2 ++
 hostapd/ctrl_iface.c  | 28 ++++++++++++++++++++++++++++
 hostapd/hostapd_cli.c | 22 ++++++++++++++++++++++
 4 files changed, 53 insertions(+), 1 deletion(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 653a0e4cc..6883277d3 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -1910,7 +1910,7 @@ static int hs20_parse_conn_capab(struct hostapd_bss_config *bss, char *buf,
 }
 
 
-static int hs20_parse_wan_metrics(struct hostapd_bss_config *bss, char *buf,
+int hs20_parse_wan_metrics(struct hostapd_bss_config *bss, char *buf,
 				  int line)
 {
 	u8 *wan_metrics;
diff --git a/hostapd/config_file.h b/hostapd/config_file.h
index c98bdb683..e033c271a 100644
--- a/hostapd/config_file.h
+++ b/hostapd/config_file.h
@@ -14,4 +14,6 @@ int hostapd_set_iface(struct hostapd_config *conf,
 		      struct hostapd_bss_config *bss, const char *field,
 		      char *value);
 
+int hs20_parse_wan_metrics(struct hostapd_bss_config *bss, char *buf, int line);
+
 #endif /* CONFIG_FILE_H */
diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index fe51d9868..25d24a755 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -5176,6 +5176,31 @@ static int hostapd_ctrl_iface_send_action_frame(struct hostapd_data *hapd, char
 	return ret;
 }
 
+static int hostapd_ctrl_iface_update_wan_metrics(struct hostapd_data *hapd,
+						const char *cmd)
+{
+	char buf[128];
+
+	hapd = mxl_get_bss_index(cmd, hapd->iface);
+	if (NULL == hapd) {
+		return -1;
+	}
+
+	cmd = os_strchr(cmd, ' ');
+	if (NULL == cmd) {
+		return -1;
+	}
+	cmd++;
+
+	os_strlcpy(buf, cmd, 128);
+	wpa_printf(MSG_DEBUG, "hostapd_ctrl_iface_update_wan_metrics, bss=%s", hapd->conf->iface);
+
+	if (hs20_parse_wan_metrics(hapd->conf, buf, 0) != 0)
+		return -1;
+
+	return 0;
+}
+
 static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 					      char *buf, char *reply,
 					      int reply_size,
@@ -5816,6 +5841,9 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "SEND_ACTION_FRAME ", sizeof("SEND_ACTION_FRAME ") - 1) == 0) {
 		if (hostapd_ctrl_iface_send_action_frame(hapd, buf + sizeof("SEND_ACTION_FRAME ") - 1))
 			reply_len = -1;
+	} else if (os_strncmp(buf, "UPDATE_WAN_METRICS ", 19) == 0) {
+		if (hostapd_ctrl_iface_update_wan_metrics(hapd, buf + 19))
+			reply_len = -1;
 	} else {
 		os_memcpy(reply, "UNKNOWN COMMAND\n", 16);
 		reply_len = 16;
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 367b282a4..c097fac95 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -2157,6 +2157,26 @@ static int hostapd_cli_cmd_send_action_frame(struct wpa_ctrl *ctrl, int argc, ch
 	return hostapd_cli_cmd(ctrl, "SEND_ACTION_FRAME", 4, argc, argv);
 }
 
+static int hostapd_cli_cmd_update_wan_metrics(struct wpa_ctrl *ctrl,
+				int argc, char *argv[])
+{
+	char buf[200];
+	int res;
+
+	if (argc != 2) {
+		printf("Invalid 'update_wan_metrics' command - "
+			   "two arguments are needed\n");
+		return -1;
+	}
+
+	res = sprintf_s(buf, sizeof(buf), "UPDATE_WAN_METRICS %s %s", argv[0], argv[1]);
+	if (res <= 0) {
+		printf("%s: sprintf_s error %d", __func__, res);
+		return -1;
+	}
+	return wpa_ctrl_command(ctrl, buf);
+}
+
 struct hostapd_cli_cmd {
 	const char *cmd;
 	int (*handler)(struct wpa_ctrl *ctrl, int argc, char *argv[]);
@@ -2434,6 +2454,8 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 	  "= get STAs key ID and passphrase if set in wpa_psk_file" },
 	{ "send_action_frame", hostapd_cli_cmd_send_action_frame, NULL,
 	  "<bss> <mac_addr> <freq> frame=<frame_buf>" },
+	{ "update_wan_metrics", hostapd_cli_cmd_update_wan_metrics, NULL,
+	  "<WAN Info>:<DL Speed>:<UL Speed>:<DL Load>:<UL Load>:<LMD> = update_wan_metrics" },
 	{ NULL, NULL, NULL, NULL }
 };
 
-- 
2.43.0

